Merge of bdist devel branch to master

mcyprian opened the pull request:
Following changes were performed: 
- Implementation of metadata extractor based on distutils command
- Redesign of previous extractors and 
- A few other small fixes

Commit:
Redesign metadata extractors
- Turn LocalMetadataExtractor into abstract class
- Move setup.py specific methods to SetupPyMetadataExtractor
- Add pypi_metadata_extension decorator to extract metadata from PyPi

Commit:
Fix local name and version getter to process wheel archives properly

Commit:
Add concept of DistMetadataExtractor

Commit:
Improve DistMetadataExtractor
- rename command to extract_distribution
- redirect stdout and stderr while extracting data
- use WheelMetadataExtractor if local file is wheel

Commit:
Prefer zip archive if tar.gz is missing, fixes #32
- use wrapper instead of monkey_patch for zip archives
  to fix extractall method

Commit:
Add unit tests for DistMetadataExtractor

Commit:
Use pytest to run tests

Commit:
Use extract_distribution class attribute to store distribution object

Commit:
Fix aborting redirection of streams to logger
- redirect stdout to logging debug level

hroncok committed:
Do not return None for classifiers, but [] instead

Fixes #34

Commit:
Add missing files to MANIFEST.in, add LICENSE to test_data dir

Commit:
Add extraction of metadata of package without classifiers test

Commit:
Add option -s to create spec file, fixes #35

hroncok commented on pyp2rpm/archive.py:
```diff
@@ -117,6 +133,8 @@ def extractor_cls(self):
         return file_cls
 
+
+
```
Why?

mcyprian replied:
I probably forgot to remove extra lines or I have added them accidentally wile editing previous functions, I will fix it.

hroncok commented on pyp2rpm/archive.py:
```diff
@@ -139,6 +157,21 @@ def get_content_of_file(self, name, full_path=False):  # TODO: log if file can't
+        """
+        if self.handle:
+            for member in self.handle.getmembers():
+                if (full_path and member.name == name)\
```
See http://stackoverflow.com/questions/181530/python-style-multiple-line-conditions-in-ifs

mcyprian replied:
I always tried to find out how to do this without using \ thanks.

hroncok commented on pyp2rpm/convertor.py:
```diff
@@ -157,28 +157,26 @@ def metadata_extractor(self):
+                self.version,
+                self.rpm_name,
+                self.venv,
+               self.base_python_version)
```
Wouldn't it be cleaner to save the class to variable a and call the variable? Something like:

```
if ...:
    extractor_cls = metadata_extractors.WheelMetadataExtractor
else:
    extractor_cls = metadata_extractors.DistMetadataExtractor
self._metadata_extractor = extractor_cls(
    self.local_file,
    self.name,
    self.name_convertor,
    self.version,
    self.rpm_name,
    self.venv,
    self.base_python_version)
```

There is also inconsistent usage of spaces/indent (lines 169 and 179).

mcyprian replied:
Seems like a good idea.

hroncok commented on pyp2rpm/extract_distribution.py:
```diff
@@ -0,0 +1,76 @@
+            raise DistutilsOptionError("{} is not a list".format(var))
+        return var 
+
+distutils.command.bdist_rpm.bdist_rpm = extract_distribution
```
How does this work?

mcyprian replied:
extracted setup.py is run like this in DistMetadataExtractor's **init** method:
`extract_distribution.run_setup(setup_py, 'bdist_rpm')`

extract_distribution command is executed instead of bdist_rpm thanks to this  assignment.
There is other way to do it using --command-packages option: https://docs.python.org/2/distutils/extending.html , but this doesn't work for me, I am not sure why.

hroncok commented on pyp2rpm/extract_distribution.py:
```diff
@@ -0,0 +1,76 @@
+        self.distribution.conflicts = [dep.replace('!=', '=') 
+                                      for dep in self.distribution.run_requires if '!=' in dep]
+
+        if (getattr(self.distribution, 'entry_points', None)
```
I know why is this happening, but maybe add a comment for others?

hroncok commented on pyp2rpm/extract_distribution.py:
```diff
@@ -0,0 +1,76 @@
+    def _list(var):
+        if var is None:
+            return []
+        elif not isinstance(var, list):
```
Are you OK with this not accepting generators, tuples, etc? I Don't know where the values are coming form, but I wold try this instead:

```
try:
    return list(var)
except TypeError:
    raise Whatever...
```

But that might provide bad results on strings :(

mcyprian replied:
Method is used in this module to convert various kinds of requires to the correct form, var parameter is always list or None, I think this implementation works well here, list(None) is probably not what we want.

hroncok commented on pyp2rpm/logger.py:
```diff
@@ -9,6 +9,21 @@
+
+    def write(self, message):
+        if message not in ('\n', ''):
+            if message[-1] == '\n':
```
maybe try `.rstrip('\n')` instead?

hroncok commented on pyp2rpm/metadata_extractors.py:
```diff
@@ -3,25 +3,89 @@
+        url = ''
+        md5_digest = None
+    
+        if len(release_urls):
```
This entire thing is very hard to read, care to explain what's happening?

mcyprian replied:
Do you mean the data update from failure_dict? Method set_from(dict, update=True) will append specific key-value data from dict dictionary to data object only if the key is not already there after previous extractions.

hroncok commented on pyp2rpm/metadata_extractors.py:
```diff
@@ -3,25 +3,89 @@
 
     """Base class for metadata extractors"""
 
+    __metaclass__ = ABCMeta
```
Comment explaining why?

hroncok commented on pyp2rpm/metadata_extractors.py:
```diff
@@ -38,6 +103,72 @@ def name_convert_deps_list(self, deps_list):
+                                              self.base_python_version)
+            return extractor.get_venv_data
+        except VirtualenvFailException:
+             logger.error("Skipping virtualenv metadata extraction")
```
I would add "...because of error"

It would be also very kind for the user to see what error happened.

hroncok commented on pyp2rpm/metadata_extractors.py:
```diff
@@ -219,167 +332,131 @@ def data_from_archive(self):
+                    setup_py = glob.glob(temp_dir + "/{0}*/".format(self.name) + 'setup.py')[0]
+                except IndexError:
+                    print("setup.py not found; maybe local_file is not proper source archive")
+                    logger.error("setup.py not found, metadata extraction failed.")
```
print and log?

mcyprian replied:
Exiting the tool with error code and without any message to user doesn't seem to be correct behaviour (maybe it should be printed to stderr instead). And log would be incomplete without information like this.What solution do you propose?

hroncok replied:
Print to stderr.

hroncok commented on pyp2rpm/package_data.py:
```diff
@@ -37,7 +37,7 @@ def __getattr__(self, name):
         elif name == 'changelog_date_packager':
             return self.get_changelog_date_packager()
-        return self.data.get(name, 'TODO:')
+        return self.data.get(name, "TODO:")
```
why?

hroncok commented:
You don't follow PEP8 very much:

```
$ git diff master..bdist_devel | python3-pep8 --diff
./pyp2rpm/archive.py:136:17: E122 continuation line missing indentation or outdented
./pyp2rpm/archive.py:138:80: E501 line too long (81 > 79 characters)
./pyp2rpm/archive.py:172:80: E501 line too long (98 > 79 characters)
./pyp2rpm/archive.py:175:80: E501 line too long (88 > 79 characters)
./pyp2rpm/bin.py:41:80: E501 line too long (99 > 79 characters)
./pyp2rpm/bin.py:68:80: E501 line too long (94 > 79 characters)
./pyp2rpm/convertor.py:79:1: W293 blank line contains whitespace
./pyp2rpm/convertor.py:158:80: E501 line too long (85 > 79 characters)
./pyp2rpm/convertor.py:163:80: E501 line too long (85 > 79 characters)
./pyp2rpm/convertor.py:174:80: E501 line too long (86 > 79 characters)
./pyp2rpm/extract_distribution.py:1:1: E902 FileNotFoundError: [Errno 2] No such file or directory: './pyp2rpm/extract_distribution.py'
./pyp2rpm/logger.py:9:80: E501 line too long (88 > 79 characters)
./pyp2rpm/logger.py:21:80: E501 line too long (81 > 79 characters)
./pyp2rpm/metadata_extractors.py:9:80: E501 line too long (80 > 79 characters)
./pyp2rpm/metadata_extractors.py:32:33: E225 missing whitespace around operator
./pyp2rpm/metadata_extractors.py:44:80: E501 line too long (86 > 79 characters)
./pyp2rpm/metadata_extractors.py:52:80: E501 line too long (94 > 79 characters)
./pyp2rpm/metadata_extractors.py:53:50: E261 at least two spaces before inline comment
./pyp2rpm/metadata_extractors.py:55:80: E501 line too long (96 > 79 characters)
./pyp2rpm/metadata_extractors.py:71:80: E501 line too long (98 > 79 characters)
./pyp2rpm/metadata_extractors.py:77:80: E501 line too long (91 > 79 characters)
./pyp2rpm/metadata_extractors.py:79:80: E501 line too long (86 > 79 characters)
./pyp2rpm/metadata_extractors.py:87:80: E501 line too long (91 > 79 characters)
./pyp2rpm/metadata_extractors.py:98:80: E501 line too long (85 > 79 characters)
./pyp2rpm/metadata_extractors.py:127:80: E501 line too long (80 > 79 characters)
./pyp2rpm/metadata_extractors.py:139:80: E501 line too long (98 > 79 characters)
./pyp2rpm/metadata_extractors.py:147:80: E501 line too long (98 > 79 characters)
./pyp2rpm/metadata_extractors.py:164:80: E501 line too long (87 > 79 characters)
./pyp2rpm/metadata_extractors.py:168:80: E501 line too long (113 > 79 characters)
./pyp2rpm/metadata_extractors.py:173:80: E501 line too long (92 > 79 characters)
./pyp2rpm/metadata_extractors.py:230:42: E127 continuation line over-indented for visual indent
./pyp2rpm/metadata_extractors.py:361:1: W293 blank line contains whitespace
./pyp2rpm/metadata_extractors.py:364:80: E501 line too long (87 > 79 characters)
./pyp2rpm/metadata_extractors.py:365:1: W293 blank line contains whitespace
./pyp2rpm/metadata_extractors.py:366:80: E501 line too long (84 > 79 characters)
./pyp2rpm/metadata_extractors.py:370:80: E501 line too long (83 > 79 characters)
./pyp2rpm/metadata_extractors.py:371:80: E501 line too long (88 > 79 characters)
./pyp2rpm/metadata_extractors.py:377:80: E501 line too long (80 > 79 characters)
./pyp2rpm/metadata_extractors.py:392:80: E501 line too long (90 > 79 characters)
./pyp2rpm/metadata_extractors.py:414:80: E501 line too long (95 > 79 characters)
./pyp2rpm/metadata_extractors.py:445:38: E127 continuation line over-indented for visual indent
./pyp2rpm/package_getters.py:39:80: E501 line too long (96 > 79 characters)
./pyp2rpm/package_getters.py:78:80: W291 trailing whitespace
./pyp2rpm/package_getters.py:81:5: E303 too many blank lines (2)
./pyp2rpm/package_getters.py:112:80: E501 line too long (91 > 79 characters)
./pyp2rpm/package_getters.py:114:80: E501 line too long (82 > 79 characters)
./pyp2rpm/package_getters.py:115:80: E501 line too long (114 > 79 characters)
./pyp2rpm/package_getters.py:117:80: E501 line too long (89 > 79 characters)
./pyp2rpm/package_getters.py:124:80: E501 line too long (88 > 79 characters)
./pyp2rpm/package_getters.py:128:80: E501 line too long (93 > 79 characters)
./pyp2rpm/package_getters.py:130:80: E501 line too long (88 > 79 characters)
./pyp2rpm/package_getters.py:150:80: E501 line too long (85 > 79 characters)
./pyp2rpm/settings.py:11:80: E501 line too long (96 > 79 characters)
./pyp2rpm/settings.py:16:80: E501 line too long (81 > 79 characters)
./pyp2rpm/settings.py:17:80: E501 line too long (82 > 79 characters)
./pyp2rpm/settings.py:18:80: E501 line too long (96 > 79 characters)
./pyp2rpm/settings.py:19:80: E501 line too long (101 > 79 characters)
./pyp2rpm/settings.py:20:80: E501 line too long (94 > 79 characters)
./pyp2rpm/settings.py:21:80: E501 line too long (84 > 79 characters)
./pyp2rpm/settings.py:24:80: E501 line too long (93 > 79 characters)
./pyp2rpm/settings.py:25:80: E501 line too long (104 > 79 characters)
./pyp2rpm/settings.py:26:80: E501 line too long (104 > 79 characters)
./pyp2rpm/utils.py:56:1: E302 expected 2 blank lines, found 1
./pyp2rpm/utils.py:110:1: E302 expected 2 blank lines, found 1
./pyp2rpm/utils.py:115:1: E302 expected 2 blank lines, found 1
./tests/test_convertor.py:12:1: E302 expected 2 blank lines, found 1
./tests/test_convertor.py:13:39: E251 unexpected spaces around keyword / parameter equals
./tests/test_convertor.py:13:41: E251 unexpected spaces around keyword / parameter equals
./tests/test_convertor.py:13:80: E501 line too long (81 > 79 characters)
./tests/test_convertor.py:14:31: E261 at least two spaces before inline comment
./tests/test_convertor.py:38:30: E251 unexpected spaces around keyword / parameter equals
./tests/test_convertor.py:38:32: E251 unexpected spaces around keyword / parameter equals
./tests/test_metadata_extractors.py:17:1: E302 expected 2 blank lines, found 1
./tests/test_metadata_extractors.py:25:80: E501 line too long (112 > 79 characters)
./tests/test_metadata_extractors.py:26:80: E501 line too long (111 > 79 characters)
./tests/test_metadata_extractors.py:27:80: E501 line too long (118 > 79 characters)
./tests/test_metadata_extractors.py:28:80: E501 line too long (118 > 79 characters)
./tests/test_metadata_extractors.py:29:80: E501 line too long (119 > 79 characters)
./tests/test_metadata_extractors.py:30:80: E501 line too long (127 > 79 characters)
./tests/test_metadata_extractors.py:31:18: E124 closing bracket does not match visual indentation
./tests/test_metadata_extractors.py:34:80: E501 line too long (98 > 79 characters)
./tests/test_metadata_extractors.py:35:80: E501 line too long (94 > 79 characters)
./tests/test_metadata_extractors.py:37:80: E501 line too long (85 > 79 characters)
./tests/test_metadata_extractors.py:107:21: E251 unexpected spaces around keyword / parameter equals
./tests/test_metadata_extractors.py:107:23: E251 unexpected spaces around keyword / parameter equals
./tests/test_metadata_extractors.py:107:80: E501 line too long (87 > 79 characters)
./tests/test_metadata_extractors.py:108:80: E501 line too long (115 > 79 characters)
./tests/test_metadata_extractors.py:109:21: E251 unexpected spaces around keyword / parameter equals
./tests/test_metadata_extractors.py:109:23: E251 unexpected spaces around keyword / parameter equals
./tests/test_metadata_extractors.py:110:80: E501 line too long (93 > 79 characters)
./tests/test_metadata_extractors.py:111:80: E501 line too long (86 > 79 characters)
./tests/test_metadata_extractors.py:129:15: E231 missing whitespace after ','
./tests/test_metadata_extractors.py:130:80: E501 line too long (85 > 79 characters)
./tests/test_metadata_extractors.py:138:1: E302 expected 2 blank lines, found 1
./tests/test_metadata_extractors.py:141:36: E261 at least two spaces before inline comment
./tests/test_metadata_extractors.py:143:80: E501 line too long (117 > 79 characters)
./tests/test_metadata_extractors.py:144:80: E501 line too long (117 > 79 characters)
./tests/test_metadata_extractors.py:145:18: E124 closing bracket does not match visual indentation
./tests/test_metadata_extractors.py:148:37: E261 at least two spaces before inline comment
./tests/test_metadata_extractors.py:149:22: E231 missing whitespace after ','
./tests/test_package_getters.py:40:80: E501 line too long (81 > 79 characters)
./tests/test_package_getters.py:41:80: E501 line too long (83 > 79 characters)
./tests/test_package_getters.py:42:80: E501 line too long (80 > 79 characters)
./tests/test_package_getters.py:45:18: E124 closing bracket does not match visual indentation
./tests/test_package_getters.py:74:80: E501 line too long (84 > 79 characters)
```

Commit:
Improve virtualenv module error handling

Commit:
Minor changes of module metadata_extractors
- Add abstract method data_from_archive to LocalMetadataExtractor
- print error message to stderr when setup.py is missing

Commit:
Fix issues using autopep8 and pylint tools

Commit:
Remove all temporary directories properly after the tests run

mcyprian commented:
Everything should be fixed now, I have also run tool autopep8 to change formatting of modules violating pep8.

hroncok commented:
Looks good to me. Didn't test the functions, as I trust your judgement and tests.
